tests/runtests.py

```diff
@@ -354,18 +343,29 @@ def django_tests(verbosity, interactive, failfast, keepdb, reverse,
                  test_labels, debug_sql, parallel, tags, exclude_tags,
                  test_name_patterns, start_at, start_after, pdb, buffer,
                  timing, shuffle):
-    actual_parallel = actual_test_processes(parallel)
+    if parallel in {0, 'auto'}:
+        max_parallel = get_max_test_processes()
+    else:
+        max_parallel = parallel
 
     if verbosity >= 1:
         msg = "Testing against Django installed in '%s'" % os.path.dirname(django.__file__)
-        if actual_parallel > 1:
-            msg += " with up to %d processes" % actual_parallel
+        if max_parallel > 1:
+            msg += " with up to %d processes" % max_parallel
         print(msg)
```

Comment:
Would it be better to move this log message to after where `parallel` is done being computed below? Otherwise, if `parallel` needs to be reset to `1` below, maybe that can be logged below so people know it was changed.

Reply:
`setup_run_tests()` has its own messages, so I'd leave it here :thinking: . It's still correct with "up to" in the sentence.

Reply:
I'm thinking of a case where the first message makes it look like tests are running in parallel, but the database check further down causes tests not to be run in parallel. It seems like the user might never find this out. (The "up to" seems like it'd be misleading in this case because there would be no chance of it ever getting "up to" that number, even with enough tests where the processes could be used / would be needed.)

Another option would be to put the parallel computation and initial log message in `setup_collect_tests()` after `django.setup()`:
https://github.com/django/django/blob/4fe3774c729f3fd5105b3001fe69a70bdca95ac3/tests/runtests.py#L249-L250
Then that message would still be first, and computing the parallel value could be done in one place -- it wouldn't be spread across two code blocks.

If you're uncomfortable putting that logic so deep within, it could instead be done by passing in a `post_setup` callable that `setup_collect_tests()` invokes immediately after `django.setup()`.

Reply:
I'd prefer to discuss it separately :thinking: . This patch is complicated even without extra cleanups :exploding_head:

Reply:
Okay, I could take a crack at some of that follow-up clean-up afterwards, if you want.